Notebook 1: Introduction to Statistics

1. What is Statistics?
   
Understand what statistics means and how we collect, organize, analyze and interpret data.

3. Why Statistics is Important in AI/ML
   
Understand how statistics helps with:

- Understanding datasets
- Finding patterns
- Data preprocessing
- Model evaluation
- Making decisions from data

3. Statistics vs Machine Learning
  
Understand the difference between:

- Statistics → understanding data and relationships
- Machine Learning → learning patterns from data to make predictions/decisions

4. Descriptive Statistics
Cover how we summarize existing data using things such as:
- Mean
- Median
- Mode
- Minimum
- Maximum
- Range
- Standard deviation
5. Inferential Statistics
Understand how we use a sample to draw conclusions about a larger population.
6. Population
The complete group we are interested in studying.
7. Sample
A smaller subset selected from the population.
8. Census
Collecting information from every member of the population.
9. Sampling
The process of selecting a sample from a population.
10. Variables
A characteristic or value that can change between observations.
11. Features
Input variables used by a machine-learning model to make predictions.
12. Observation
One individual record/data point in a dataset.
13. Dataset
A collection of observations and their variables/features.
14. Quantitative Data
Numerical data that can be measured or counted.
Examples:
- Age = 25
- Salary = ₹50,000
- Height = 176 cm
15. Qualitative Data
Descriptive or categorical data rather than numerical measurements.
Examples:
- Gender/category
- Product type
- City
- Customer feedback
16. Continuous Data
Numerical data that can take values within a range.
Examples:
- Height = 176.5 cm
- Temperature = 36.7°C
- Weight = 72.4 kg
17. Discrete Data
Countable numerical values, usually whole numbers.
Examples:
- Number of customers = 100
- Number of orders = 25
- Number of cars = 8
18. Structured Data
Data organized into a predefined structure, usually rows and columns.
Example:
Age	Salary	City
25	50000	Hyderabad
30	65000	Chennai


This is common in databases and CSV files.
19. Unstructured Data
Data without a fixed tabular structure.
Examples:
- Images
- Videos
- Audio
- Text documents
- Social-media posts
This is especially important in AI/ML, NLP, computer vision and Generative AI.

Notebook 2: Measures of Central Tendency


1. Mean
   
Step 1 – Understand the Concept

What is Mean?
Mean is the average value of a set of numbers.
We calculate it by adding all the values and dividing the total by the number of values.

Mean is useful when we want to represent a dataset with one central value.

For example, if five employees have salaries of ₹30,000, ₹35,000, ₹40,000, ₹45,000 and ₹50,000, the mean salary gives us the average salary of these employees.
Formula
\[
Mean = \frac{\sum x}{n}
\]
Where:
- \(x\) = each value
- \(n\) = number of values
- \(\sum x\) = sum of all values

AI/ML Example
Suppose a dataset contains customer ages:
20, 25, 30, 35, 40

The mean age can be used as a basic summary of the dataset and can help us understand the distribution of the data.

In [2]:
%pip install numpy


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.8/6.8 MB 13.0 MB/s  0:00:00 eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [3]:
import numpy as np

marks = np.array([60, 70, 80, 90, 100])

mean = np.mean(marks)

print("Mean:", mean)

Mean: 80.0


2. Median

Step 1 – Understand the Concept

What is Median?
Median is the middle value when the data is arranged in ascending or descending order.
If there is an even number of values, we take the average of the two middle values.

Median is especially useful when the dataset contains outliers because extreme values don't affect it as much as they affect the mean.

Formula

For an odd number of observations:
\[
Median = \text{middle value}
\]
For an even number:
\[
Median = \frac{\text{two middle values}}{2}
\]


In [4]:
import numpy as np

salaries = np.array([25000, 30000, 35000, 40000, 500000])

median = np.median(salaries)

print("Median:", median)

Median: 35000.0


3. Mode
   
Step 1 – Understand the Concept

What is Mode?
Mode is the value that occurs most frequently in a dataset.
Unlike mean and median, mode can be used directly with categorical data.

A dataset can have:
- One mode → unimodal
- Two modes → bimodal
- Multiple modes → multimodal
  
Formula
There isn't a mathematical calculation like mean.
The mode is simply:
\[
Mode = \text{most frequently occurring value}
\]

In [5]:
from collections import Counter

sizes = ["M", "L", "M", "S", "M", "L", "M"]

counts = Counter(sizes)

mode = counts.most_common(1)[0][0]

print("Mode:", mode)

Mode: M


Interpretation
M is the most frequently occurring shirt size.
Where is it used in AI/ML?
Mode can be useful for:
- Categorical data
- Missing categorical values
- Customer segmentation
- Survey analysis
- Understanding the most common category

4. Weighted Mean

Step 1 Understand the Concept

What is Weighted Mean?
Weighted mean is an average where some values have more importance than others.
In a normal mean, every value contributes equally.
    
In a weighted mean, each value has a weight.
    
Formula
\[
Weighted\ Mean = \frac{\sum wx}{\sum w}
\]
Where:
- \(x\) = value
- \(w\) = weight

Business Example

A company may calculate an employee's overall performance using:
- Project performance → 50%
- Technical skills → 30%
- Communication → 20%
  
The different weights represent different levels of importance.

AI/ML Example
Weighted averages can be used when combining metrics where some observations or categories should contribute more than others.

In [11]:
import numpy as np

scores = np.array([80, 70, 90])
weights = np.array([0.2, 0.3, 0.5])

weighted_mean = np.average(scores, weights=weights)

print("Weighted Mean:", weighted_mean)

Weighted Mean: 82.0


Interpretation

The final weighted score is 82.

The final exam has the highest weight, so its score has the largest effect on the result.
Where is it used in AI/ML?
Weighted means are useful in:
- Evaluation metrics
- Weighted datasets
- Business scoring systems
- Imbalanced datasets
- Combining multiple measurements

5. Geometric Mean
   
Step 1 – Understand the Concept

What is Geometric Mean?

Geometric mean is an average based on multiplication rather than addition.
It is particularly useful when working with:
- Growth rates
- Returns
- Ratios
- Multiplicative changes
  
Formula

For \(n\) values:
\[
Geometric\ Mean = \sqrt[n]{x_1x_2...x_n}
\]

In [12]:
import numpy as np

growth_factors = np.array([1.10, 1.20])

geometric_mean = np.prod(growth_factors) ** (1 / len(growth_factors))

print("Geometric Mean:", geometric_mean)
print("Average Growth Rate:", (geometric_mean - 1) * 100, "%")

Geometric Mean: 1.1489125293076057
Average Growth Rate: 14.891252930760569 %


Interpretation
The average compound growth factor is approximately 1.149, corresponding to about 14.9% growth.
Where is it used in AI/ML?
It can be useful when dealing with:
- Growth rates
- Ratios
- Multiplicative processes
- Certain performance metrics

6. Harmonic Mean
   
Step 1 – Understand the Concept

What is Harmonic Mean?

Harmonic mean is an average that is especially useful for rates and ratios.
It gives more importance to smaller values compared with the arithmetic mean.

Formula
\[
Harmonic\ Mean =
\frac{n}{\sum\frac{1}{x_i}}
\]
For two values:
\[
H = \frac{2ab}{a+b}
\]

Business Example
A company may use harmonic mean when combining performance expressed as rates, such as certain productivity or speed measurements.
AI/ML Example
One of the most important ML examples is the F1-score.
F1-score is the harmonic mean of:
- Precision
- Recall
\[
F1 = \frac{2PR}{P+R}
\]

In [13]:
import numpy as np

precision = 0.80
recall = 0.60

harmonic_mean = 2 * precision * recall / (precision + recall)

print("Harmonic Mean / F1 Score:", harmonic_mean)

Harmonic Mean / F1 Score: 0.6857142857142857


Interpretation

The F1-score is approximately 0.686.
It combines precision and recall into one value while giving importance to both.
Where is it used in AI/ML?
Harmonic mean is especially important for:
- Precision and Recall
- F1-score
- Classification problems
- Information retrieval
- Search systems

# 1. Range

## Step 1 – Understand the Concept

### What is Range?

Range is the simplest measure of dispersion.

It tells us how far the data extends from the smallest value to the largest value.

In simple words, Range tells us the difference between the maximum and minimum values.

### Formula

Range = Maximum Value - Minimum Value

### Why do we use Range?

Range gives us a quick idea about the overall spread of a dataset.

A larger range means the values cover a wider interval.

A smaller range means the values are closer together.

---

## Step 2 – Demonstrate the Concept

### Numerical Example

Consider the following marks:

10, 20, 30, 40, 50

Maximum = 50

Minimum = 10

Range:

50 - 10 = 40

Therefore, the range is 40.

### Business Example

Suppose a company's daily sales are:

₹10,000, ₹15,000, ₹18,000, ₹25,000 and ₹40,000

Range:

₹40,000 - ₹10,000 = ₹30,000

This tells the company how widely daily sales varied.

### AI/ML Example

Suppose a feature contains values from 10 to 100.

The range is 90.

This gives us a quick understanding of the scale and spread of that feature.

---

## Step 3 – Implementation

We will use NumPy to calculate the maximum, minimum and range.

### What does the program do?

The program finds the largest and smallest values and subtracts the minimum from the maximum.

### Interpretation

The resulting value represents the total spread between the smallest and largest observations.

### AI/ML Usage

Range can be useful during:

- Exploratory Data Analysis
- Feature analysis
- Data preprocessing
- Understanding feature scales

In [16]:
import numpy as np

marks = np.array([10, 20, 30, 40, 50])

maximum = np.max(marks)
minimum = np.min(marks)

data_range = maximum - minimum

print("Maximum:", maximum)
print("Minimum:", minimum)
print("Range:", data_range)

Maximum: 50
Minimum: 10
Range: 40


# 2. Variance

## Step 1 – Understand the Concept

### What is Variance?

Variance measures how far the values in a dataset are spread from the mean.

It calculates the average of the squared differences between each value and the mean.

In simple words:

Variance tells us how much the values vary around their average.

### Formula

For population variance:

Variance = Σ(x - μ)² / N

Where:

x = individual value

μ = population mean

N = total number of observations

For sample variance:

s² = Σ(x - x̄)² / (n - 1)

### Why do we square the differences?

Some values will be above the mean and some below the mean.

If we simply add the differences, positive and negative values could cancel each other.

Squaring makes all differences positive and gives more importance to larger deviations.

---

## Step 2 – Demonstrate the Concept

Consider:

2, 4, 6

### Step 1: Calculate the mean

Mean = (2 + 4 + 6) / 3

Mean = 4

### Step 2: Calculate differences from mean

2 - 4 = -2

4 - 4 = 0

6 - 4 = 2

### Step 3: Square the differences

(-2)² = 4

0² = 0

2² = 4

### Step 4: Calculate variance

Variance = (4 + 0 + 4) / 3

Variance = 2.67

Therefore, the population variance is approximately 2.67.

### Business Example

Two stores can have the same average daily sales but different variance.

A store with high variance has more unpredictable sales.

### AI/ML Example

Variance helps us understand how much a feature varies.

It is also important in:

- Statistical analysis
- Feature analysis
- Data preprocessing
- Machine learning algorithms

---

## Step 3 – Implementation

### What does the program do?

The program calculates the population variance using NumPy.

### Interpretation

A larger variance means the values are more spread out from the mean.

A smaller variance means the values are closer to the mean.

### AI/ML Usage

Variance is important when understanding feature variability and statistical distributions.

In [17]:
import numpy as np

data = np.array([2, 4, 6])

mean = np.mean(data)
variance = np.var(data)

print("Mean:", mean)
print("Population Variance:", variance)

Mean: 4.0
Population Variance: 2.6666666666666665


# 3. Standard Deviation

## Step 1 – Understand the Concept

### What is Standard Deviation?

Standard deviation tells us how much the values typically differ from the mean.

It is the square root of variance.

Because variance uses squared units, its value can sometimes be difficult to interpret.

Standard deviation brings the measurement back to the original unit of the data.

### Formula

Population Standard Deviation:

σ = √[Σ(x - μ)² / N]

### Relationship

Standard Deviation = √Variance

### Why do we use Standard Deviation?

It gives us an understandable measure of spread.

For example, if the average salary is ₹50,000 and the standard deviation is ₹5,000, we know that salaries typically vary around the mean by roughly ₹5,000.

---

## Step 2 – Demonstrate the Concept

Consider:

2, 4, 6

Mean = 4

Population variance ≈ 2.67

Therefore:

Standard Deviation = √2.67

Standard Deviation ≈ 1.63

### Business Example

A company can compare the consistency of daily sales.

If Store A has a standard deviation of ₹2,000 and Store B has ₹10,000, Store B has more variation in daily sales.

### AI/ML Example

Standard deviation is widely used in:

- Feature scaling
- Standardization
- Outlier detection
- Exploratory Data Analysis
- Statistical modeling

---

## Step 3 – Implementation

### What does the program do?

The program calculates the population standard deviation using NumPy.

### Interpretation

A higher standard deviation means greater spread.

A lower standard deviation means the values are closer to the mean.

### AI/ML Usage

Standard deviation is especially important for standardization:

z = (x - mean) / standard deviation

This converts features to a common scale.


In [18]:
import numpy as np

data = np.array([2, 4, 6])

mean = np.mean(data)
variance = np.var(data)
standard_deviation = np.std(data)

print("Mean:", mean)
print("Variance:", variance)
print("Standard Deviation:", standard_deviation)

Mean: 4.0
Variance: 2.6666666666666665
Standard Deviation: 1.632993161855452


# 4. Quartiles

## Step 1 – Understand the Concept

### What are Quartiles?

Quartiles divide an ordered dataset into four approximately equal parts.

There are three important quartiles:

Q1 = First Quartile = 25th percentile

Q2 = Second Quartile = 50th percentile = Median

Q3 = Third Quartile = 75th percentile

### Why do we use Quartiles?

Quartiles help us understand how the data is distributed.

They are especially useful when the dataset contains outliers because they are less affected by extreme values than the mean.

---

## Step 2 – Demonstrate the Concept

Consider:

10, 20, 30, 40, 50, 60, 70

The median is 40.

Q1 represents the 25th percentile.

Q2 represents the 50th percentile.

Q3 represents the 75th percentile.

In Python, NumPy can calculate these values directly.

### Business Example

A company can divide customers into four groups based on spending:

Q1 → lower-spending customers

Q2 → middle range

Q3 → higher-spending customers

Above Q3 → highest spending group

### AI/ML Example

Quartiles are useful during Exploratory Data Analysis to understand feature distributions.

They are also used when calculating the Interquartile Range.

---

## Step 3 – Implementation

### What does the program do?

The program calculates Q1, Q2 and Q3 using NumPy.

### Interpretation

Q1 tells us where approximately 25% of the observations fall below.

Q2 is the median.

Q3 tells us where approximately 75% of the observations fall below.

### AI/ML Usage

Quartiles are commonly used for:

- EDA
- Outlier detection
- Distribution analysis
- Feature understanding

In [19]:
import numpy as np

data = np.array([10, 20, 30, 40, 50, 60, 70])

q1 = np.percentile(data, 25)
q2 = np.percentile(data, 50)
q3 = np.percentile(data, 75)

print("Q1:", q1)
print("Q2 (Median):", q2)
print("Q3:", q3)

Q1: 25.0
Q2 (Median): 40.0
Q3: 55.0


# 5. Percentiles

## Step 1 – Understand the Concept

### What is a Percentile?

A percentile tells us the value below which a particular percentage of observations fall.

For example:

The 25th percentile means approximately 25% of the observations are below or equal to that value.

The 50th percentile is the median.

The 90th percentile means approximately 90% of observations are below or equal to that value.

### Why do we use Percentiles?

Percentiles help us understand the relative position of a value within a dataset.

---

## Step 2 – Demonstrate the Concept

Suppose the 90th percentile of customer spending is ₹10,000.

This means approximately 90% of customers spend ₹10,000 or less, while approximately 10% spend more.

### Business Example

An e-commerce company could use the 90th percentile of delivery time to understand a high-end delivery-time threshold.

For example, if the 90th percentile delivery time is 5 days, approximately 90% of deliveries take 5 days or less.

### AI/ML Example

Percentiles can be used for:

- Outlier detection
- Feature analysis
- Setting thresholds
- Understanding distributions
- Data preprocessing

---

## Step 3 – Implementation

### What does the program do?

The program calculates the 25th, 50th, 75th and 90th percentiles.

### Interpretation

Each percentile describes the position of a value within the dataset.

### AI/ML Usage

Percentiles are frequently used during EDA to understand the distribution and identify extreme observations.

In [20]:
import numpy as np

data = np.array([10, 20, 30, 40, 50, 60, 70, 80, 90, 100])

p25 = np.percentile(data, 25)
p50 = np.percentile(data, 50)
p75 = np.percentile(data, 75)
p90 = np.percentile(data, 90)

print("25th Percentile:", p25)
print("50th Percentile:", p50)
print("75th Percentile:", p75)
print("90th Percentile:", p90)

25th Percentile: 32.5
50th Percentile: 55.0
75th Percentile: 77.5
90th Percentile: 91.0


# 6. Interquartile Range (IQR)

## Step 1 – Understand the Concept

### What is IQR?

Interquartile Range, or IQR, measures the spread of the middle 50% of a dataset.

It is calculated using Q1 and Q3.

### Formula

IQR = Q3 - Q1

Where:

Q1 = 25th percentile

Q3 = 75th percentile

### Why do we use IQR?

IQR is useful because it is less affected by extreme values.

It is commonly used for detecting potential outliers.

### Outlier Detection

A common method uses:

Lower Bound = Q1 - 1.5 × IQR

Upper Bound = Q3 + 1.5 × IQR

Values outside these boundaries can be considered potential outliers.

---

## Step 2 – Demonstrate the Concept

Consider:

10, 20, 30, 40, 50, 60, 70, 100

Suppose:

Q1 = 25

Q3 = 62.5

Then:

IQR = 62.5 - 25

IQR = 37.5

### Business Example

Suppose employee salaries contain a few extremely high salaries.

IQR can help the company understand the spread of the middle 50% without being heavily influenced by those extreme salaries.

### AI/ML Example

IQR is commonly used in data preprocessing to identify potential outliers before training a machine learning model.

---

## Step 3 – Implementation

### What does the program do?

The program calculates Q1, Q3 and IQR.

It also calculates the lower and upper bounds used for potential outlier detection.

### Interpretation

Observations outside the calculated bounds can be investigated as potential outliers.

They are not automatically errors. We should understand the data before removing them.

### AI/ML Usage

IQR can be useful for:

- Outlier detection
- Data cleaning
- EDA
- Feature preprocessing

In [21]:
import numpy as np

data = np.array([10, 20, 30, 40, 50, 60, 70, 100])

q1 = np.percentile(data, 25)
q3 = np.percentile(data, 75)

iqr = q3 - q1

lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)

print("Q1:", q1)
print("Q3:", q3)
print("IQR:", iqr)
print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)

outliers = data[(data < lower_bound) | (data > upper_bound)]

print("Potential Outliers:", outliers)

Q1: 27.5
Q3: 62.5
IQR: 35.0
Lower Bound: -25.0
Upper Bound: 115.0
Potential Outliers: []


# 7. Coefficient of Variation

## Step 1 – Understand the Concept

### What is Coefficient of Variation?

Coefficient of Variation, or CV, measures the standard deviation relative to the mean.

It is usually expressed as a percentage.

It helps us compare variability between datasets even when their means or scales are different.

### Formula

CV = (Standard Deviation / Mean) × 100

### Why do we use CV?

Standard deviation alone does not always allow us to compare variability between datasets with different averages.

CV expresses the variation relative to the mean.

---

## Step 2 – Demonstrate the Concept

Suppose:

Mean = 100

Standard Deviation = 20

CV:

(20 / 100) × 100 = 20%

Therefore, the coefficient of variation is 20%.

### Business Example

Suppose two investment options have different average returns and different standard deviations.

CV can help compare their relative variability.

### AI/ML Example

CV can be used to compare the relative variability of numerical features.

However, it should be used carefully when the mean is zero or very close to zero because the calculation can become unstable or meaningless.

---

## Step 3 – Implementation

### What does the program do?

The program calculates the mean, standard deviation and coefficient of variation.

### Interpretation

A higher CV means greater variability relative to the mean.

A lower CV means the data has less variability relative to the mean.

### AI/ML Usage

CV can help with:

- Comparing variability
- Feature analysis
- Business analytics
- Statistical analysis

In [23]:
import numpy as np

data = np.array([80, 90, 100, 110, 120])

mean = np.mean(data)
standard_deviation = np.std(data)

coefficient_of_variation = (standard_deviation / mean) * 100

print("Mean:", mean)
print("Standard Deviation:", standard_deviation)
print("Coefficient of Variation:", coefficient_of_variation, "%")

Mean: 100.0
Standard Deviation: 14.142135623730951
Coefficient of Variation: 14.142135623730951 %
